ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/TransactionReaper.java

```diff
@@ -714,23 +714,27 @@ private final void shutdown(boolean waitForTransactions)
         synchronized (this) {
             _inShutdown = true;
 
-            /*
-                * If the caller does not want to wait for the normal transaction timeout
-                * periods to elapse before terminating, then we first start by enabling
-                * our time machine!
-                */
-
-            if (!waitForTransactions) {
-                _reaperElements.setAllTimeoutsToZero();
-            }
-
             /*
                 * Wait for all of the transactions to
                 * terminate normally.
                 */
             while (!_reaperElements.isEmpty()) {
+                /*
+                 * If the caller does not want to wait for the normal transaction timeout
+                 * periods to elapse before terminating, then we first start by enabling
+                 * our time machine!
+                 */
+
+                if (!waitForTransactions) {
+                    _reaperElements.setAllTimeoutsToZero();
```

Comment:
Please can I ask why this block was moved into the `!_reaperElements.isEmpty()` condition?

Reply:
Because there are no guarntees on the handling of reaper elements added simultanously or after shutdown is called.

Reply:
Well there is a comment https://github.com/jbosstm/narayana/blob/main/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/TransactionReaper.java#L929-L931 which is in the the public documentation of the API so hopefully client applications would not be adding things in the situations you mentioned (I want to thank @jmfinelli who mentioned something in this area)?

Reply:
I'll leave it up to you guys if you want to forego the added safety of moving this into the loop. If it gets moved back out, I'll re-examine the quakus logic - at first glance it didn't seem to ensure that a transaction could not be started concurrently with shutdown due to the async nature of the events in play.

Reply:
I think Quarkus should be made to respect this requirement if there is indeed a problem with that. From my reading of the API we should not need to change the location of this code. I will raise a pull request on your branch @shawkins. With that moved back, personally I would be ready to approve the pull request (assuming it passes CI).

Reply:
I raised https://github.com/shawkins/narayana/pull/1/ to try to get this change into your JBTM-3830 branch pre-completing the review /cc @mmusgrov @jmfinelli